Allow flags to comment TOAs when writing to file

scottransom opened the pull request:
This is for discussion with #1011, but it seems useful and has tests.
Any TOA with an `ignore` or `cut` flag will be commented on write.

aarchiba commented on tests/test_toa_writer.py:
```diff
@@ -42,10 +42,14 @@ def test_roundtrip_topo_toa_Tempo2format(self):
         ts.write_TOA_file("testtopo.tim", format="Tempo2")
         ts2 = toa.get_TOAs("testtopo.tim")
         print(ts.table, ts2.table)
+        assert ts2.ntoas == 1  # one should be commented
```
Can you make this a separate test, so we can tell which things are broken and which are not?

scottransom replied:
Done

aarchiba commented on src/pint/toa.py:
```diff
@@ -1728,11 +1728,12 @@ def write_TOA_file(self, filename, name="unk", format="tempo2"):
             else:
                 toatime_out = toatime
-            out_str = format_toa_line(
+            out_str = "C " if ("ignore" in flags or "cut" in flags) else ""
```
Can you make the "ignore" flag an optional argument to `write_TOA_flag`? So that users don't find their TOAs disappearing unexpectedly, and so that users *can* write files that include `-cut` TOAs that aren't commented out?

scottransom replied:
How will people find their TOAs disappearing unexpectedly given that 1) they have to explicitly add these flags (or they would have had to do something non-standard to the .tim file (i.e. uncommenting them) before reading them in) and 2) they don't disappear, they are just commented?  The whole point of this change is to make a simple way that we can select against TOAs that we don't want to fit or whatever, and which we can easily flag to be commented *if* we re-write the TOAs.

I'm not seeing an advantage to an optional argument (which users have to remember) -- especially since if "-cut" means that we've cut something, we really don't want it included!  If you want to make a note about a TOA, use a different flag.

scottransom replied:
If we want to make an *additional* way of commenting TOAs via an optional argument which provides a mask array, I suppose that is fine.  The utility and ability to use flags for this type of thing is exactly why TOAs have flags now.

aarchiba replied:
My thought was that if a user has a tim file in which some TOAs have a `-cut` flag that they did not add, they may be surprised that those TOAs are commented out when the TOAs are written out, and they will have disappeared when the user tries to read the tim file back in. 

I also anticipate wanting to write out a tim file with `-cut` flags indicating reasons for cutting TOAs but with those TOAs not actually commented out, so that this tim file can easily be used to evaluate our excision criteria. So I'd like to be able to turn off the commenting feature.

scottransom replied:
This was my point at the beginning of the discussion in #1011.  We are trying to use comments and/or the `-cut` flag to do things which they (IMO) shouldn't be doing.  NANOGrav started using the `-cut` flag to describe *why* we cut (thereby commenting out) a TOA.  But the bottom line is that `-cut` == commented as defined by NANOGrav.

So for your first point, the user shouldn't have TOAs that have a cut flag and are un-commented unless they (or someone else) modified them.   But even if they do, they still need to do the non-standard thing of writing out a new TOA file *and* be ignorant as to what the flags contain and mean in order to be "surprised".

If you (or someone else) wants to evaluate *possible* reasons to cut TOAs and test them (i.e. your 2nd point), then you should be using a different flag since all strings are valid.  But `-cut` means that we have cut the TOA!  Please don't change  the definition of `-cut`.

scottransom commented:
I added a warning (should it be INFO instead?) that TOAs with `-ignore` and/or `-cut` flags will be commented.  And I also added an optional argument for commenting to be turned off.  I think the default should be to comment, though, given that those flags explicitly mean that we are ignoring those TOAs.

aarchiba commented:
Does anyone actually use `-ignore` or is it something you're introducing?

scottransom commented:
Introducing.  As an example of a way to add a flag that means that we want to ignore a TOA and comment it out.  Could be something else.  But, for example, I think `pintk` should have a way to select TOAs for commenting or ignoring (perhaps temporarily;  and adding a flag like this could do it).  I didn't use `-comment`, because that flag sounds like you could be giving a standard comment *about* a TOA.

aarchiba commented:
I ask because the original purpose for this was to allow `-cut` to work, so we are presumably going to make that produce comments. Do we need to introduce a second flag to do the same thing?

scottransom commented:
We certainly don't need to.  But for interactive use, I don't think `-cut` is very intuitive.  I used `-ignore` as an example of an alias that might be a bit more intuitive.

swiggumj commented:
Previous NANOGrav releases have used `-cut` for this purpose and in timing analysis config files, we have an `ignore` block. I guess I don't mind one way or the other, but I suggested `-cut` from a backwards-compatibility standpoint.

scottransom commented:
Yeah, I think we definitely want `-cut` for this.  The question is simply if we want any other aliases for commenting TOAs using flags.  I personally think `-ignore` makes sense for interactive work.  But I don't feel strongly.

scottransom commented:
I don't know of any others, besides what TEMPO2 has defined (and which are described in the T2 Users Manual):
```
-r (receiver)
-b (backend)
-i/-info (info)
-dm
-p (phase)
-t (telescope)
```
And IPTA DR1 paper (Verbiest et al) mentioned using a `-simul` flag and a `-dmo` flag (simultaneous observations and DM offset, respectively).

paulray commented:
Looks like you are missing `-pn` for pulse numbers

aarchiba commented on src/pint/toa.py:
```diff
@@ -1708,6 +1711,11 @@ def write_TOA_file(self, filename, name="unk", format="tempo2"):
+        # Warn the user if any of the TOAs have -ignore or -cut flags that
+        # they will be commented in the output file
+        if isinstance(commentflag, str) and len(self.get_flag_value(commentflag)[1]):
+            log.warning(f"TOAs with '-{commentflag}' flags will be commented")
```
We already have too many messages. If you are worried about the user being surprised by commented TOAs, make the default None; if you are not worried, leave the warning out.

scottransom replied:
I'm fine leaving the warning out.  I put it in there specifically because you said that you were worried that people might be surprised by having TOAs commented.

aarchiba replied:
Indeed; my preferred solution is to default to not commenting any TOAs. But I don't think a warning is a good idea.

aarchiba commented on src/pint/toa.py:
```diff
@@ -1728,11 +1736,14 @@ def write_TOA_file(self, filename, name="unk", format="tempo2"):
                 toatime_out = toatime
-            out_str = format_toa_line(
+            out_str = (
+                "C " if isinstance(commentflag, str) and (commentflag in flags) else ""
```
None is hashable, so you can just leave this as `commentflag in flags` and rely on the keys all being actual flags.

scottransom replied:
What I'm doing is stricter than that.  I'm requiring that `commentflag` must be a string.  If it is anything else, including a number, and list, an array, or a `None`, then it won't write a comment.

aarchiba commented on tests/test_toa_writer.py:
```diff
@@ -49,6 +49,25 @@ def test_roundtrip_topo_toa_Tempo2format(self):
+        t2 = toa.TOA(t1time, obs="gbt", freq=0.0)
+        ts = toa.get_TOAs_list([t1, t2], ephem="DE421")
+        assert ts.ntoas == 2
+        ts.table[1]["flags"]["ignore"] = True  # flag one TOA
```
Maybe also check the default behaviour? That is, if the default is "cut", maybe check that that happens when you pass no argument?

scottransom replied:
I'm fine doing that, but checking for `ignore` tests for the default behavior (since `ignore` is a string like `cut`, just not the default -- and I don't think we need to test that Python's optional arguments work) as well as the user choosing something besides the default string.  Seems like a waste of a (small amount) of test time.

aarchiba replied:
We had a bug earlier this year where the test suite never executed a method with the default value of its argument, so that default value had the wrong units and we never noticed. Tests also detect when the public API changes, and the default value is part of that public API, so it's a good idea to check that.

aarchiba commented on tests/test_toa_writer.py:
```diff
@@ -49,6 +49,25 @@ def test_roundtrip_topo_toa_Tempo2format(self):
+        ts.write_TOA_file("testtopo.tim", format="Tempo2", commentflag="ignore")
+        ts4 = toa.get_TOAs("testtopo.tim")
+        assert ts4.ntoas == 1  # one should be commented
+        print(ts.table, ts2.table, ts3.table, ts4.table)
```
This will never do anything - `pytest` only prints the output if an exception arose, in which case this line is never executed.

scottransom replied:
Those print() functions were in each of the class methods from before.  I was just carrying them along.  (and note that they *are* executed and actually print stuff if you use the `-s` option for pytest (which I use a lot for debugging tests).

aarchiba commented on tests/test_toa_writer.py:
```diff
@@ -45,34 +43,54 @@ def test_roundtrip_topo_toa_Tempo2format(self):
+        t2 = toa.TOA(t1time, obs="gbt", freq=0.0)
+        ts = toa.get_TOAs_list([t1, t2], ephem="DE421")
+        assert ts.ntoas == 2
+        ts.write_TOA_file("testtopo.tim", format="Tempo2")
```
Does this leave a stray `testtopo.tim` lying around? Can you use a temporary directory or a StringIO?

scottransom replied:
Yes, it leaves a stray testtopo.tim lying around (and the other tests in there also leave files).  I can clean that up, although that is really beyond the scope of what the pull request is trying to do.

scottransom replied:
Done.  And changed all into pytest format.

aarchiba commented on tests/test_toa_writer.py:
```diff
@@ -45,34 +43,54 @@ def test_roundtrip_topo_toa_Tempo2format(self):
+        ts.write_TOA_file("testtopo.tim", format="Tempo2")
+        ts2 = toa.get_TOAs("testtopo.tim")
+        assert ts2.ntoas == 2  # none should be commented
+        ts.table[0]["flags"]["cut"] = "do not like"  # default comment flag
```
What happens when we use flags with spaces in them? Does PINT handle this correctly? (What would correct handling look like - there isn't a defined form of quoting is there?)

scottransom replied:
Yeah, I don't even know what that should look like.  Especially if a flag (or the flag's value) with spaces is in a .tim file.

scottransom replied:
I changed the value to have s/ /_/ in this case.

scottransom commented:
This should be good to go now.
